extensions/parquet/table/src/main/java/io/deephaven/parquet/table/util/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,273 @@
+        // null context equivalent to clearing the context
+        if (context != null && !(context instanceof S3ChannelContext)) {
+            throw new IllegalArgumentException("context must be null or an instance of ChannelContext, provided context " +
+                    " of class " + context.getClass().getName());
+        }
+        this.context = (S3ChannelContext) context;
+    }
+
+    @Override
+    public int read(@NotNull final ByteBuffer destination) throws ClosedChannelException {
+        Assert.neqNull(context, "context");
+        if (!destination.hasRemaining()) {
+            return 0;
+        }
+        final long localPosition = position;
+        checkClosed(localPosition);
+        if (localPosition >= size) {
+            // We are finished reading
+            return -1;
+        }
+
+        final int currFragmentIndex = fragmentIndexForByteNumber(localPosition);
+        final ByteBuffer currentFragment;
+        if (currFragmentIndex == finalFragmentIdx && finalFragment != null) {
+            currentFragment = finalFragment.asReadOnlyBuffer();
+        } else {
+            // Send async read request the current fragment, if it's not already in the cache
+            final S3ChannelContext.FragmentContext fragmentContext = context.getFragmentContext(currFragmentIndex);
+            final CompletableFuture<ByteBuffer> fetchCurrFragment;
+            if (fragmentContext != null && fragmentContext.fragmentIndex == currFragmentIndex) {
+                fetchCurrFragment = fragmentContext.future;
+            } else {
+                fetchCurrFragment = computeFragmentFuture(currFragmentIndex);
+                context.setFragmentContext(currFragmentIndex, fetchCurrFragment);
+            }
+
+            // Send async requests for read-ahead buffers and store them in the cache
+            final int numFragmentsToLoad = Math.min(readAheadCount, numFragmentsInObject - currFragmentIndex - 1);
+            for (int i = 0; i < numFragmentsToLoad; i++) {
+                final int readAheadFragmentIndex = i + currFragmentIndex + 1;
+                final S3ChannelContext.FragmentContext readAheadFragmentContext = context.getFragmentContext(readAheadFragmentIndex);
+                if (readAheadFragmentContext == null || readAheadFragmentContext.fragmentIndex != readAheadFragmentIndex) {
+                    context.setFragmentContext(readAheadFragmentIndex, computeFragmentFuture(readAheadFragmentIndex));
+                }
+            }
+
+            // Wait till the current fragment is fetched
+            try {
+                currentFragment = fetchCurrFragment.get(timeout, timeUnit).asReadOnlyBuffer();
+            } catch (final InterruptedException | ExecutionException | TimeoutException | RuntimeException e) {
+                throw new UncheckedDeephavenException("Failed to fetch fragment " + currFragmentIndex + " for file "
+                        + key + " in S3 bucket " + bucket, e);
```

Comment:
Ultimately, there's an interface mismatch that I think we'll eventually want to address. POSIX is typically fast and shallow queue, while BLOB is typically slow/async and deep queued. We can't work around it right now since we are dealing with blocking `java.nio.channels.SeekableByteChannel#read`, but I do want to at least think about how the engine could take advantage of async IO with better interfaces eventually.